client/src/main/java/com/aws/greengrass/cli/adapter/impl/NucleusAdapterIpcClientImpl.java

```diff
@@ -209,6 +229,222 @@ public CreateDebugPasswordResponse createDebugPassword() {
+            }
+
+            // To stop subscribing, close the stream.
+            responseHandler.closeStream();
+        } catch (InterruptedException e) {
+            System.out.println("IPC interrupted.");
+        } catch (ExecutionException e) {
+            System.err.println("Exception occurred when using IPC.");
+            e.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    public static class SubscriptionMqttResponseHandler implements StreamResponseHandler<IoTCoreMessage> {
+
+        @Override
+        public void onStreamEvent(IoTCoreMessage ioTCoreMessage) {
+            try {
+                String message = new String(ioTCoreMessage.getMessage().getPayload(),
+                        StandardCharsets.UTF_8);
+                System.out.printf("Received new message : %s%n",  message);
+            } catch (Exception e) {
+                System.err.println("Exception occurred while processing subscription response " +
+                        "message.");
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public boolean onStreamError(Throwable error) {
+            System.err.println("Received a stream error.");
+            error.printStackTrace();
+            return false;
+        }
+
+        @Override
+        public void onStreamClosed() {
+            System.out.println("Subscribe to IoT Core stream closed.");
+        }
+    }
+
+    private SubscribeToTopicResponseHandler subscribeToTopic(GreengrassCoreIPCClient greengrassCoreIPCClient
+            , String topic, StreamResponseHandler<SubscriptionResponseMessage> streamResponseHandler) {
+        SubscribeToTopicRequest subscribeToTopicRequest = new SubscribeToTopicRequest();
+        subscribeToTopicRequest.setTopic(topic);
+        return greengrassCoreIPCClient.subscribeToTopic(subscribeToTopicRequest,
+                Optional.of(streamResponseHandler));
+    }
+
+
+    public static class SubscriptionResponseHandler implements StreamResponseHandler<SubscriptionResponseMessage> {
+
+        private final String topic;
+        private static final ObjectMapper SERIALIZER = new ObjectMapper();
+
+        public SubscriptionResponseHandler(String topic) {
+            this.topic = topic;
+        }
+
+        @Override
+        public void onStreamEvent(SubscriptionResponseMessage subscriptionResponseMessage) {
+            try {
+                Optional<Map<String, Object>> jsonMessage = Optional.empty();
+                if (subscriptionResponseMessage.getJsonMessage() != null) {
+                    jsonMessage = Optional.of(subscriptionResponseMessage.getJsonMessage().getMessage());
+                }
+                Optional<byte[]> binaryMessage = Optional.empty();
+                if (subscriptionResponseMessage.getBinaryMessage() != null) {
+                    binaryMessage = Optional.of(subscriptionResponseMessage.getBinaryMessage().getMessage());
```

Comment:
optional shouldn't be needed here. Just use the if/else. 

When you use optional, always use `ofNullable`, not `of`.